### PR TITLE
Rewrite the filters.

### PR DESCRIPTION
**Pull request by kaikai1024**
### Description of the Change

* Add skeleton.
* Implement new_block_filter.
* Implement new_filter.
* Implement uninstall.
* Implement get_filter_logs.
* Implement get_filter_changes.
* Add prune according the lifetime.

TODO: Rewrite `bloom_possibilities` function.

### CODE REVIEW

File: cita-chain/core/src/filters/rpc_filter.rs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+use cita_types::H256;
+use jsonrpc_types::rpc_types::{Filter, FilterChanges, Log};
+
+/// The RPC interfaces about filter.
+///     * newFilter
+///     * newBlockFilter
+///     * getFilterChanges
+///     * getFilterLogs
+///     * uninstallFilter
+/// *Not include `getLogs`*.
```
Comment (ouwenkg) [inline cita-chain/core/src/filters/rpc_filter.rs]: Why

File: cita-chain/core/src/filters/rpc_filter.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    // Get the logs for the filter with the given id.
+    // https://docs.citahub.com/zh-CN/cita/rpc-guide/rpc#getfilterlogs
+    fn get_filter_logs(&self, id: usize) -> Option<Vec<Log>>;
+    // Remove the filter with the given id.
+    // https://docs.citahub.com/zh-CN/cita/rpc-guide/rpc#uninstallfilter
+    fn uninstall_filter(&self, id: usize) -> bool;
+}
+
+trait FilterHelper {
+    // Get the block for the filter with the given id. Used by other interfaces.
```
Comment (ouwenkg) [inline cita-chain/core/src/filters/rpc_filter.rs]: Get block number

File: cita-chain/core/src/filters/rpc_filter.rs (hunk 1 above)
Comment (kaikai1024) [reply cita-chain/core/src/filters/rpc_filter.rs]: it just call `get_logs` of  `Chain`

File: cita-chain/core/src/filters/rpc_filter.rs (hunk 2 above)
Comment (kaikai1024) [reply cita-chain/core/src/filters/rpc_filter.rs]: yeah. I will check all the codes later.

File: cita-chain/core/src/filters/rpc_filter.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        };
+
+        // Update the block filter
+        filterdb
+            .try_lock()
+            .unwrap()
+            .new_block_filter(id, current_number + 1);
+        drop(filterdb);
+        changes
+    }
```
Comment (ouwenkg) [inline cita-chain/core/src/filters/rpc_filter.rs]: Why new such a block filter?

Comment (kaikai1024) [reply cita-chain/core/src/filters/rpc_filter.rs]: Check the comment.
It's a bad name, i will refactor all the names later.
It's used to record the current block when call the `getFIlterChanges`.

File: cita-chain/src/forward.rs [hunk 4]
```diff
@@ -21,7 +21,7 @@ use std::sync::atomic::Ordering;
 use std::sync::Arc;
 
 use cita_types::H256;
-use core::filters::eth_filter::EthFilter;
+use core::filters::rpc_filter::RpcFilter as FilterMethod;
```
Comment (ouwenkg) [inline cita-chain/src/forward.rs]: Where `FilterMethod` used?

Comment (kaikai1024) [reply cita-chain/src/forward.rs]: Check the error message:
```
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
            `use core::filters::rpc_filter::RpcFilter;`
```

**Comment by ouwenkg** [comment]
LGTM

**Pull request merged**